datastore/prometheus.go

```diff
@@ -0,0 +1,318 @@
+	}
+}
+
+func (c *eventCache) get(uid string) (Event, bool) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	val, ok := c.c[uid]
+	return val, ok
+}
+
+func (c *eventCache) set(uid string, e Event) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.c[uid] = e
+}
+
+func NewPrometheusExporter(ctx context.Context) *PrometheusExporter {
+	exporter := &PrometheusExporter{
+		ctx:           ctx,
+		reg:           prometheus.NewRegistry(),
+		podCache:      newEventCache(),
+		svcCache:      newEventCache(),
+		reqChanBuffer: make(chan Request, 10000),
+		podEventChan:  make(chan Event, 100),
+		svcEventChan:  make(chan Event, 100),
+	}
+
+	exporter.latencyHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "alaz",
+			Name:      "http_latency",
+			Buckets:   []float64{0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000},
+		},
+		[]string{"fromPod", "fromNamespace", "toPod", "toNamespace", "toHost", "toPort", "toService",
+			"fromAcornProject", "fromAcornApp", "fromAcornContainer", "fromAcornAccountId",
+			"toAcornProject", "toAcornApp", "toAcornContainer", "toAcornAccountId"},
+	)
+	exporter.reg.MustRegister(exporter.latencyHistogram)
+
+	exporter.statusCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "alaz",
+			Name:      "http_status",
+		},
+		[]string{"fromPod", "fromNamespace", "toPod", "toNamespace", "toHost", "toPort", "toService", "status",
+			"fromAcornProject", "fromAcornApp", "fromAcornContainer", "fromAcornAccountId",
+			"toAcornProject", "toAcornApp", "toAcornContainer", "toAcornAccountId"},
+	)
+	exporter.reg.MustRegister(exporter.statusCounter)
+
+	// Set up caches
+	go exporter.startCache(exporter.podCache, exporter.podEventChan)
+	go exporter.startCache(exporter.svcCache, exporter.svcEventChan)
+	go exporter.handleReqs()
+
+	go func() {
+		http.Handle("/metricz", promhttp.HandlerFor(exporter.reg, promhttp.HandlerOpts{}))
+		go func() {
+			if err := http.ListenAndServe(":8080", nil); err != nil {
+				log.Logger.Error().Err(err).Msg("error while serving metrics")
+			}
+		}()
+		<-exporter.ctx.Done()
+		log.Logger.Info().Msg("Prometheus exporter stopped")
+	}()
+
+	return exporter
+}
+
+func (p *PrometheusExporter) startCache(cache *eventCache, ch chan Event) {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case obj := <-ch:
+			cache.set(obj.GetUID(), obj)
+		}
+	}
+}
+
+func (p *PrometheusExporter) handleReqs() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case req := <-p.reqChanBuffer:
+			p.handleReq(req)
+		}
+	}
+}
+
+func (p *PrometheusExporter) handleReq(req Request) {
+	fromPod, found := p.podCache.get(req.FromUID)
+	if !found {
+		log.Logger.Warn().Str("uid", req.FromUID).Msg("pod not found")
+		return
+	}
+
+	switch req.ToType {
+	case "pod":
+		toPod, found := p.podCache.get(req.ToUID)
+		if !found {
+			return
+		}
+		p.updateMetricsToPod(fromPod.(*PodEvent), toPod.(*PodEvent), req)
+	case "service":
```

Comment:
In all of my testing, I have yet to see the `service` case happen (even when a pod is contacting another pod via a service hostname), but I figured it would be best to handle it anyway.

Reply:
We no longer need to distinguish this.